wave_matching/src/icp.cpp

```diff
@@ -75,4 +79,367 @@ bool ICPMatcher::match() {
+            MM (2, 5) -= corrs_aver[ci] (0);
+            MM (3, 4) -= corrs_aver[ci] (0) * corrs_aver[ci] (2);
+            MM (3, 5) -= corrs_aver[ci] (0) * corrs_aver[ci] (1);
+            MM (4, 5) -= corrs_aver[ci] (1) * corrs_aver[ci] (2);
+            MM (3, 3) += corrs_aver[ci] (1) * corrs_aver[ci] (1) + corrs_aver[ci] (2) * corrs_aver[ci] (2);
+            MM (4, 4) += corrs_aver[ci] (0) * corrs_aver[ci] (0) + corrs_aver[ci] (1) * corrs_aver[ci] (1);
+            MM (5, 5) += corrs_aver[ci] (0) * corrs_aver[ci] (0) + corrs_aver[ci] (2) * corrs_aver[ci] (2);
+
+            // Fast computation of M'Z
+            MZ (0) += corrs_diff[ci] (0);
+            MZ (1) += corrs_diff[ci] (1);
+            MZ (2) += corrs_diff[ci] (2);
+            MZ (3) += corrs_aver[ci] (1) * corrs_diff[ci] (2) - corrs_aver[ci] (2) * corrs_diff[ci] (1);
+            MZ (4) += corrs_aver[ci] (0) * corrs_diff[ci] (1) - corrs_aver[ci] (1) * corrs_diff[ci] (0);
+            MZ (5) += corrs_aver[ci] (2) * corrs_diff[ci] (0) - corrs_aver[ci] (0) * corrs_diff[ci] (2);
+
+        }
+        // Remaining elements of M'M
+        MM (0, 0) = MM (1, 1) = MM (2, 2) = static_cast<float> (numCorr);
+        MM (4, 0) = MM (0, 4);
+        MM (5, 0) = MM (0, 5);
+        MM (3, 1) = MM (1, 3);
+        MM (4, 1) = MM (1, 4);
+        MM (3, 2) = MM (2, 3);
+        MM (5, 2) = MM (2, 5);
+        MM (4, 3) = MM (3, 4);
+        MM (5, 3) = MM (3, 5);
+        MM (5, 4) = MM (4, 5);
+
+        // Compute pose difference estimation
+        Vec6 D = static_cast<Vec6> (MM.inverse () * MZ);
+
+        // Compute s^2
+        float ss = 0.0f;
+        for (int ci = 0; ci != numCorr; ++ci)  // ci = correspondence iterator
+        {
+            ss += static_cast<float> (pow (corrs_diff[ci] (0) - (D (0) + corrs_aver[ci] (2) * D (5) - corrs_aver[ci] (1) * D (4)), 2.0f)
+                                      + pow (corrs_diff[ci] (1) - (D (1) + corrs_aver[ci] (0) * D (4) - corrs_aver[ci] (2) * D (3)), 2.0f)
+                                      + pow (corrs_diff[ci] (2) - (D (2) + corrs_aver[ci] (1) * D (3) - corrs_aver[ci] (0) * D (5)), 2.0f));
+        }
+
+        // When reaching the limitations of computation due to linearization
+        if (ss < 0.0000000000001 || !pcl_isfinite (ss))
+        {
+            this->information = Mat6::Identity();
+            return;
+        }
+
+        this->information = MM * (1.0f/ss);
+    }
+}
+
+// This is an implementation of the Haralick or Censi covariance approximation
+// for ICP
+// The core idea behind this is that the covariance of the cost f'n J wrt
+// optimization variable x is
+// cov(x) ~= (d2J/dx2)^-1*(d2J/dzdx)*cov(z)*(d2J/dzdx)'*(d2J/dx2)^-1
+
+// Idea was taken from
+// https://censi.science/pub/research/2007-icra-icpcov.pdf
+
+// This is an implementation for euler angles, what is below is a cleaned up
+// version of
+// http://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=7153246
```

Comment:
The whole matcher actually needs to be updated with doxygen docs. I'll do that in a seperate PR.